Define seed and expanded key format and introduce secret key identifier for ML-* (#171)
openpgp-pqc/draft-openpgp-pqc: draft-ietf-openpgp-pqc.md

 - A fixed-length octet string containing the ML-KEM public key, whose length depends on the algorithm ID as specified in {{tab-mlkem-artifacts}}.

The algorithm-specific secret key is this series of values:

 - A fixed-length octet string of the encoded secret scalar, whose encoding and length depend on the algorithm ID as specified in {{ecc-kem}}.

 - A one-octet ML-KEM private key format specifier defined in {{mlkem-privkey-spec}}.

>  - An octet string containing the ML-KEM secret key.
>    The length depends on the private key format specifier in the previous field.

Reviewer: ```suggestion
 - An octet string containing the ML-KEM secret key, whose length depends on the private key format identifier in the previous field.
```
Little closer to the previous sentence.

> #### ML-KEM Private Key Format Specifier {#mlkem-privkey-spec}

Reviewer: ```suggestion
#### ML-KEM Private Key Format Identifier {#mlkem-privkey-spec}
```

> {: title="ML-KEM Private Key Format Specifier" #tab-mlkem-privkey-spec}

Reviewer: ```suggestion
{: title="ML-KEM Private Key Format Identifier" #tab-mlkem-privkey-spec}
```

Octet Value | Name  | Private Key Size
-----:| ------------ | ----------
0x00  | Seed Format  | 64

> This document only defines one value for the ML-KEM private key format specifier which is the 64 byte long seed format.

Reviewer: I think we could remove this sentence as it is clear from the table.

> This document only defines one value for the ML-KEM private key format specifier which is the 64 byte long seed format.

Reviewer: ```suggestion
```
I think we could remove this sentence as it is clear from the table.

 - A fixed-length octet string containing the ML-KEM public key, whose length depends on the algorithm ID as specified in {{tab-mlkem-artifacts}}.

The algorithm-specific secret key is this series of values:

 - A fixed-length octet string of the encoded secret scalar, whose encoding and length depend on the algorithm ID as specified in {{ecc-kem}}.

 - A one-octet ML-KEM secret key format identifier defined in {{mlkem-seckey-id}}.

 - An octet string containing the ML-KEM secret key, whose length depends on the secret key format identifier in the previous field.

#### ML-KEM Secret Key Format Identifier {#mlkem-seckey-id}

{: title="ML-KEM Secret Key Format Identifier" #tab-mlkem-seckey-id}
ID | Name
-----:| ------------
0x00  | Seed Format

> 0x01  | Expanded Key Format

Reviewer: ```suggestion
0x01  | Expanded Format
```
I thin `Key` is superfluous here.

The seed format is defined in accordance with [FIPS-203], Section 3.3.
Namely, the secret key is given by the concatenation of the values of `d`  and `z`, generated in steps 1 and 2 of `ML-KEM.KeyGen` [FIPS-203], each of a length of 32 octets.
Upon parsing the seed format, or before using the secret key, for the expansion of the key, the function `ML-KEM.KeyGen_internal` [FIPS-203] has to be invoked with the parsed values of `d` and `z` as input.

The expanded key format is the value `dk` that is generated by `ML-KEM.KeyGen` [FIPS-203].

The length of the different formats is given in {{tab-mlkem-sec-size}}.

{: title="ML-KEM secret key length in octets depending on secret key format identifier" #tab-mlkem-sec-size}

> ML-KEM       | Seed Format | Expanded Key Format

Reviewer: ```suggestion
ML-KEM       | Seed Format | Expanded Format
```

------------:| ------------ | ---------
ML-KEM-768   | 64           | 2400
ML-KEM-1024  | 64           | 3168

> Implementations MUST be able to parse the seed format and MAY implement parsing of the expanded key format.

Reviewer 2: What is missing is a specification of the requirements pertaining to *writing* a key. My suggestion is
- "should create private keys in the seed format"
- "and may create them in the expanded format"

The composite ML-DSA + EdDSA schemes MUST be used only with v6 keys, as defined in [RFC9580], or newer versions defined by updates of that document.

The algorithm-specific public key is this series of values:

 - A fixed-length octet string representing the EdDSA public key, whose length depends on the algorithm ID as specified in {{tab-eddsa-artifacts}}.

 - A fixed-length octet string containing the ML-DSA public key, whose length depends on the algorithm ID as specified in {{tab-mldsa-artifacts}}.

The algorithm-specific secret key is this series of values:

 - A fixed-length octet string representing the EdDSA secret key, whose length depends on the algorithm ID as specified in {{tab-eddsa-artifacts}}.

>  - A one-octet ML-DSA private key format specifier defined in {{mldsa-privkey-spec}}.

Reviewer: ```suggestion
 - A one-octet ML-DSA private key format identifier defined in {{mldsa-privkey-spec}}.
```

>  - An octet string containing the ML-DSA secret key.
>    The length depends on the private key format specifier in the previous field.

Reviewer: ```suggestion
 - An octet string containing the ML-DSA secret key, whose length depends on the private key format identifier in the previous field.
```

> #### ML-DSA Private Key Format Specifier {#mldsa-privkey-spec}

Reviewer: ```suggestion
#### ML-DSA Private Key Format Identifier {#mldsa-privkey-spec}
```

> {: title="ML-DSA Private Key Format Specifier" #tab-mldsa-privkey-spec}

Reviewer: ```suggestion
{: title="ML-DSA Private Key Format Identifier" #tab-mldsa-privkey-spec}
```

Octet Value | Name  | Private Key Size
-----:| ------------ | ----------
0x00  | Seed Format  | 32

> This document only defines one value for the ML-DSA private key format specifier which is the 32 byte long seed format.

Reviewer: ```suggestion
```
Again, I think we could remove this sentence.

The composite ML-DSA + EdDSA schemes MUST be used only with v6 keys, as defined in [RFC9580], or newer versions defined by updates of that document.

The algorithm-specific public key is this series of values:

 - A fixed-length octet string representing the EdDSA public key, whose length depends on the algorithm ID as specified in {{tab-eddsa-artifacts}}.

 - A fixed-length octet string containing the ML-DSA public key, whose length depends on the algorithm ID as specified in {{tab-mldsa-artifacts}}.

The algorithm-specific secret key is this series of values:

 - A fixed-length octet string representing the EdDSA secret key, whose length depends on the algorithm ID as specified in {{tab-eddsa-artifacts}}.

 - A one-octet ML-DSA secret key format identifier defined in {{mldsa-seckey-id}}.

 - An octet string containing the ML-DSA secret key, whose length depends on the secret key format identifier in the previous field.

#### ML-DSA Secret Key Format Identifier {#mldsa-seckey-id}

{: title="ML-DSA Secret Key Format Identifier" #tab-mldsa-seckey-id}
ID | Name
-----:| ------------
0x00  | Seed Format

> 0x01  | Expanded Key Format

Reviewer: ```suggestion
0x01  | Expanded Format
```

The seed format is defined in accordance with [FIPS-204], Section 3.6.3.
Namely, the secret key is given by the value `xi` generated in step 1 of `ML-DSA.KeyGen` [FIPS-204].
Upon parsing the seed format, or before using the secret key, for the expansion of the key, the function `ML-DSA.KeyGen_internal` [FIPS-204] has to be invoked with the parsed value of `xi` as input.

The expanded key format is the value `sk` that is generated by `ML-DSA.KeyGen` [FIPS-204].

The length of the different formats is given in {{tab-mldsa-sec-size}}.

{: title="ML-DSA secret key length in octets depending on secret key format identifier" #tab-mldsa-sec-size}

> ML-KEM       | Seed Format | Expanded Key Format

Reviewer: ```suggestion
ML-KEM       | Seed Format | Expanded Format
```

# Notes on Algorithms

## ML-KEM and ML-DSA Secret Key Formats

> This document specifies both the seed format and the expanded secret key format for ML-KEM and ML-DSA secret keys.

Reviewer: ```suggestion
This document specifies both the seed and expanded format for ML-KEM and ML-DSA secret keys.
```

It should be noted that it is always possible to generate the expanded secret key format from the seed format, but it is not possible the other way around.
Therefore, implementations SHOULD choose the seed format whenever possible.

> Implementations MAY export a secret key using the expanded key format if this is required for interoperability with another implementation or hardware device.

Reviewer 2: I would keep all requirements regarding reading and writing secret keys to together. So far, requirements reading and writing and in three different places.

Author: Since the key format is tied to the algorithm, I think we have to address it in the two algorithm specific sections. And since this part that you replied to is relevant for both, I chose to move it here in order to not duplicate.

Do you have an idea how I should keep it all together?

Reviewer 2: Let's wait with this PR for now. Currently we assume that we stick to the seed-only format and thus won't need this one.

- Drop CCA-conversion of X25519/X448.
- Switch to hedged variant also for SLH-DSA.

## draft-ietf-openpgp-pqc-08
- Assign code points 35 and 36 for ML-KEM algorithms.

> - Add a private key format specifier for ML-DSA and ML-KEM private keys.

Reviewer: ```suggestion
- Add a private key format identifier for ML-DSA and ML-KEM private keys.
```
